website/source/docs/connect/mesh_gateway.html.md

```diff
@@ -0,0 +1,169 @@
+---
+layout: "docs"
+page_title: "Connect - Mesh Gateways"
+sidebar_current: "docs-connect-meshgateways"
+description: |-
+  A Mesh Gateway enables better routing of a Connect service's data to upstreams in other datacenters. This section details how to use Envoy and describes how you can plug in a gateway of your choice.
+---
+
+# Mesh Gateways
+
+Mesh gateways enable routing of Connect traffic between different Consul datacenters. Those datacenters
+can reside in different clouds or runtime environments where general interconnectivity between all services
+in all datacenters isn't feasible. These gateways operate by sniffing the SNI header out of the Connect session
+and then route the connection to the appropriate destination based on the server name requested. The data
+within the Connect session is not decrypted by the Gateway.
+
+![Mesh Gateway Architecture](/assets/images/mesh-gateways.png)
+
+## Prerequisites
+
+Each mesh gateway needs three things:
+
+1. A local Consul agent to manage its configuration.
+2. General network connectivity to all services within its local Consul datacenter.
+3. General network connectivity to all mesh gateways within remote Consul datacenters.
+
+## Modes of Operation
+
+Each upstream of a Connect proxy can be configured to be routed through a mesh gateway. Depending on
+your network, the proxy's connection to the gateway can happen in one of the following modes:
+
+* `local` - In this mode the Connect proxy makes its outbound connection to a gateway running in the
+  same datacenter. That gateway is then responsible for ensuring the data gets forwarded along to
+  gateways in the destination datacenter. This is the mode of operation depicted in the diagram at
+  the beginning of the page.
+
+* `remote` - In this mode the Connect proxy makes its outbound connection to a gateway running in the
+  destination datacenter. That gateway will then forward the data to the final destination service.
+
+* `none` - In this mode, no gateway is used and a Connect proxy makes its outbound connections directly
+  to the destination services.
+
+## Mesh Gateway Configuration
+
+Mesh gateways are defined very similarly to other typical services. The one exception is that a mesh gateway
+service definition may contain a `Proxy.Config` entry just like a Connect proxy service to define opaque
+configuration parameters useful for the actual proxy software.
+
+## Connect Proxy Configuration
+
+Configuring a Connect Proxy to use gateways is as simple as setting its mode of operation. This can be done
+in several different places allowing for global to more fine grained control. If the gateway mode is configured
+in multiple locations the order of precedence is as follows
+
+1. Upstream Definition
+2. Service Instance Definition
+3. Centralized `service-defaults` configuration entry
+4. Centralized `proxy-defaults` configuration entry.
+
+### Enabling Gateways Globally
```

Comment:
I think these examples may go well as new sections under the ["proxy registration" section](/docs/connect/registration.html).  We should either add them there or rename that section to indicate that it's only for Connect sidecar proxies.

Reply:
Actually, on second thought, I think these are fine here. I think service instance is the only one that might go under proxy registration since it isn't a configuration entry?

Reply:
There definitely is a little overlap as you can configure the mesh gateway mode for individual proxies. I documented the field in that section too. I was thinking that having one place to show the 4 different ways it could be configured would be useful (and mentioning the precedence)